pkg/lib/manifest2po

```diff
@@ -0,0 +1,114 @@
+"MIME-Version: 1.0\\n"
+"Content-Type: text/plain; charset=UTF-8\\n"
+"Content-Transfer-Encoding: 8bit\\n"
+"X-Generator: Cockpit manifest2po\\n"
+"""
+
+
+def get_docs_strings(docs: List[Dict[str, str]]) -> Iterable[str]:
+    for doc in docs:
+        assert 'label' in doc, 'doc entry without label'
+        yield doc['label']
+
+
+def get_keyword_strings(keywords: List[Dict[str, List[str]]]) -> Iterable[str]:
+    for keyword in keywords:
+        assert 'matches' in keyword, 'keywords entry without matches'
+        for match in keyword['matches']:
+            yield match
+
+
+def get_menu_strings(menu: Dict[str, Any]) -> Iterable[str]:
+    for _, entry in menu.items():
+        if 'label' in entry:
+            yield entry['label']
+        if 'keywords' in entry:
+            yield from get_keyword_strings(entry['keywords'])
+        if 'docs' in entry:
+            yield from get_docs_strings(entry['docs'])
+
+
+def get_bridges_strings(bridges: List[Dict[str, str]]) -> Iterable[str]:
+    for bridge in bridges:
+        if 'label' in bridge:
+            yield bridge['label']
+
+
+def get_manifest_strings(manifest: Dict[str, Any]) -> Iterable[str]:
+    if 'menu' in manifest:
+        yield from get_menu_strings(manifest['menu'])
+    if 'tools' in manifest:
+        yield from get_menu_strings(manifest['tools'])
+    if 'bridges' in manifest:
+        yield from get_bridges_strings(manifest['bridges'])
+
+
+def main() -> None:
+    parser = argparse.ArgumentParser(prog='manifest2po',
+                                     description='Extracts translatable strings from manifest.json files')
+    parser.add_argument('-d', '--directory', help='Base directory for input files')
+    parser.add_argument('-o', '--output', help='Output files', required=True)
+    parser.add_argument('files', nargs='+', help='One or more input files', type=pathlib.Path, metavar='FILE')
+
+    args = parser.parse_args()
+    strings: DefaultDict[str, Set[str]] = collections.defaultdict(set)
+
+    file: pathlib.Path
+    for file in args.files:
+        if file.name != 'manifest.json':
+            continue
+
+        # Qualify the filename if necessary
+        full_path = args.directory / file if args.directory else file
+        with open(full_path, 'r') as fp:
+            data = fp.read()
+
+            # There are variables which when not substituted can cause JSON.parse to fail
+            # Dummy replace them. None variable is going to be translated anyway
+            safe_data = re.sub(r"@.+?@", "1", data)
+            manifest = json.loads(safe_data)
+            for msgid in get_manifest_strings(manifest):
+                strings[msgid].add(str(file))
+
+    # Write PO file
+    with open(args.output, "w") as fp:
+        fp.write(PO_HEADER)
+        for msgid in strings:
+            manifest_filenames = ' '.join([f'{fname}:0' for fname in strings[msgid]])
+            fp.write(f"""
+#: {manifest_filenames}
+msgid "{msgid}"
+msgstr ""
+""")
```

Comment:
I feel like it might be easier to read if you just print the lines one at a time.  Definitely a matter of taste, though.

Reply:
I had that before, the convert to an f-string :-)